nx/lib/nx/lin_alg.ex

```diff
@@ -1759,4 +1759,72 @@ defmodule Nx.LinAlg do
     |> Nx.take_diagonal(offset: offset)
     |> Nx.product(axes: [rank - 2])
   end
+
+  @doc """
+  Return matrix rank of input M × N matrix using Singular Value Decomposition method.
+
+  Approximate the number of linearly independent rows by calculating the number
+  of singular values greater than `eps * max(singular values) * max(M, N)`.
+
+  This also appears in Numerical recipes in the discussion of SVD solutions for
+  linear least squares [1].
+
+  [1] W. H. Press, S. A. Teukolsky, W. T. Vetterling and B. P. Flannery,
+  “Numerical Recipes (3rd edition)”, Cambridge University Press, 2007, page 795.
+
+  ## Options
+    * `:eps` - Rounding error threshold used to assume values as 0. Defaults to `1.0e-10`
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 2], [3, 4]]))
+      #Nx.Tensor<
+        u64
+        2
+      >
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 1, 1, 1], [1, 1, 1, 1], [1, 2, 3, 4]]))
+      #Nx.Tensor<
+        u64
+        2
+      >
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 1, 1], [2, 2, 2], [8, 9, 10], [-2, 1, 5]]))
+      #Nx.Tensor<
+        u64
+        3
+      >
+
+  ## Error cases
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([1, 2, 3]))
+      ** (ArgumentError) tensor must have at least rank 2, got rank 1 with shape {3}
+  """
+  @doc from_backend: false
+  def matrix_rank(a, opts \\ []) do
+    # TODO: support batching when SVD supports it too
+    opts = keyword!(opts, eps: 1.0e-10)
+    shape = Nx.shape(a)
+    :ok = Nx.Shape.matrix_rank(shape)
+
+    # Calculate max dimension
+    {row_dim, col_dim} = shape
+    max_dim = if row_dim > col_dim, do: row_dim, else: col_dim
+
+    # Calculate max singular value
+    {_u, s, _v} = svd(a)
+
+    s_max =
+      s
+      |> Nx.reduce_max()
+      |> Nx.to_number()
+
+    # Set tolerance values
+    tol = opts[:eps] * max_dim * s_max
+
+    # Calculate matrix rank
+    s
+    |> Nx.greater(tol)
```

Comment:
```suggestion
    s_max = Nx.reduce_max(s)

    # Set tolerance values
    tol = Nx.multiply(opts[:eps] * max_dim, s_max)

    # Calculate matrix rank
    s
    |> Nx.greater(tol)
```

Reply:
Why we should use `Nx.multiply` instead of * here?

Reply:
If we can convert to defn we shouldn't :)

But as it stands, we need to remove the Nx.to_number call and this in turn necessitates Nx.multiply instead of *